### PR TITLE
Improve DistributionInterface

### LINKED ISSUE #12720: Generalize and protect Distribution objects
## Rationale
(1) The DistributionInterface returns a non-const reference to Distribution objects, this allows 
users to change the state of an object that might be used by another. 

(2) The Distribution objects should include pdf/cdf/quantile methods that accept inputs (e.g., mean, standard deviation) so that they can be used as a generic function, if needed.

## Description
see Rationale.

## Impact

These changes will remove the ability for users to change the state of Distribution objects from outside, which will increase stability.

### PR DESCRIPTION
**Pull request by aeslaughter**
closes #12720

### CODE REVIEW

File: modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+[]
+
+[Variables]
+  [u]
+  []
+[]
+
+[Distributions]
+  [this_is_the_wrong_type]
+    type = UniformDistribution
```
Comment (jiangwen84) [inline modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i]: should be ``TruncatedUniformDistribution``?

File: modules/stochastic_tools/include/distributions/TruncatedNormalDistribution.h [hunk 2]
```diff
@@ -25,9 +25,25 @@ class TruncatedNormalDistribution : public NormalDistribution
...
-  virtual Real quantile(const Real & p) override;
+  virtual Real pdf(const Real & x) const override;
+  virtual Real cdf(const Real & x) const override;
+  virtual Real quantile(const Real & p) const override;
+
+  Real pdf(const Real & x,
+           const Real & mean,
+           const Real & std_dev,
+           const Real & lower_bound,
+           const Real & upper_bound) const;
```
Comment (permcody) [inline modules/stochastic_tools/include/distributions/TruncatedNormalDistribution.h]: I'm curious, are you actually sharing memory with these references? If not you might consider making these just Real parameters. Sometimes copies are actually faster than references due to locality. No big deal, just a thought.

File: modules/stochastic_tools/test/include/utils/TestDistributionDirectPostprocessor.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef TESTDISTRIBUTIONDIRECTPOSTPROCESSOR_H
+#define TESTDISTRIBUTIONDIRECTPOSTPROCESSOR_H
+
+#include "GeneralPostprocessor.h"
+
+class TestDistributionDirectPostprocessor;
```
Comment (jiangwen84) [inline modules/stochastic_tools/test/include/utils/TestDistributionDirectPostprocessor.h]: suggest -> ``TestNormalDistributionDirectPostprocessor``

**Review by permcody — DISMISSED** [review]

File: modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i (hunk 1 above)
Comment (aeslaughter) [reply modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i]: It doesn't matter, it just needs to be something that is not a normal distribution to test the error message.

**Comment by permcody** [comment]
Approved from my standpoint - you may want to address Wen's comments.

File: modules/stochastic_tools/test/include/utils/TestDistributionDirectPostprocessor.h (hunk 3 above)
Comment (aeslaughter) [reply modules/stochastic_tools/test/include/utils/TestDistributionDirectPostprocessor.h]: In the future we might want to add tests for interfaces for different distributions, this class will work for that.

Comment (jiangwen84) [reply modules/stochastic_tools/test/include/utils/TestDistributionDirectPostprocessor.h]: 👍

File: modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i (hunk 1 above)
Comment (jiangwen84) [reply modules/stochastic_tools/test/tests/distributions/normal_direct_type_error.i]: My bad. I thought you had ``NormalDistribution`` there.

**Comment by jiangwen84** [comment]
Looks good to me :-)

**Review by jiangwen84 — DISMISSED** [review]

**Comment by aeslaughter** [comment]
I will merge this soon, it requires a fix to Grizzly as well.

**Review by permcody — DISMISSED** [review]

**Comment by aeslaughter** [comment]
@jiangwen84 I added the JohnsonSB and Logistic distributions, can you review those commits. I have a change ready in Grizzly as well, but I need to get this merged to next first.

File: modules/stochastic_tools/include/distributions/TruncatedNormalDistribution.h (hunk 2 above)
Comment (aeslaughter) [reply modules/stochastic_tools/include/distributions/TruncatedNormalDistribution.h]: These could be copies.

File: modules/stochastic_tools/include/distributions/LogisticDistribution.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef GRIZZLYLOGISTICDISTRIBUTION_H
```
Comment (jiangwen84) [inline modules/stochastic_tools/include/distributions/LogisticDistribution.h]: remove GRIZZLY

File: modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+           const Real & alpha_2) const;
+  Real quantile(const Real & p,
+                const Real & a,
+                const Real & b,
+                const Real & alpha_1,
+                const Real & alpha_2) const;
+
+protected:
+  /// The lower location parameter, a
+  const Real & _a;
```
Comment (jiangwen84) [inline modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h]: _a has been used as vector<Real> in ``NormalDistribution``. I am not sure if it is a bad practice.

File: modules/stochastic_tools/include/distributions/LogisticDistribution.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+#include "Distribution.h"
+
+class LogisticDistribution;
+
+template <>
+InputParameters validParams<LogisticDistribution>();
+
+/**
+ * A class used to generate a logistic distribution
+ * This provides a method to set the parameters after it has been constructed.
```
Comment (jiangwen84) [inline modules/stochastic_tools/include/distributions/LogisticDistribution.h]: delete this line

File: modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+#include "NormalDistribution.h"
+
+class JohnsonSBDistribution;
+
+template <>
+InputParameters validParams<JohnsonSBDistribution>();
+
+/**
+ * A class used to generate a Johnson SB distribution
+ * This provides a method to set the parameters after it has been constructed.
```
Comment (jiangwen84) [inline modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h]: delete this line

File: modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h (hunk 5 above)
Comment (jiangwen84) [reply modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h]: and _b

**Review by jiangwen84 — CHANGES_REQUESTED** [review]

File: modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h (hunk 5 above)
Comment (aeslaughter) [reply modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h]: good catch

**Review by jiangwen84 — APPROVED** [review]

**Comment by jiangwen84** [comment]
👍

**Review by jiangwen84 — APPROVED** [review]

File: modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+           const Real & alpha_1,
+           const Real & alpha_2) const;
+  Real quantile(const Real & p,
+                const Real & a,
+                const Real & b,
+                const Real & alpha_1,
+                const Real & alpha_2) const;
+
+protected:
+  /// The lower location parameter, a
```
Comment (jiangwen84) [inline modules/stochastic_tools/include/distributions/JohnsonSBDistribution.h]: just  minor suggestion, a - > lower and b -> upper

**Review by permcody — APPROVED** [review]

**Pull request merged**